noir_stdlib/src/hash/sha256.nr

```diff
@@ -202,66 +246,209 @@ fn verify_msg_block<let N: u32>(
+    msg_block: MSG_BLOCK,
+    mut msg_byte_ptr: BLOCK_BYTE_PTR,
+    max_int_byte_ptr: u32,
+) {
     // This variable is used to get around the compiler under-constrained check giving a warning.
     // We want to check against a constant zero, but if it does not come from the circuit inputs
     // or return values the compiler check will issue a warning.
     let zero = msg_block[0] - msg_block[0];
 
-    for i in 0..BLOCK_SIZE {
-        if i >= msg_byte_ptr {
+    // First integer which is supposed to be (partially) zero.
+    let mut int_byte_ptr = msg_byte_ptr / INT_SIZE;
+
+    // Check partial zeroes.
+    let modulo = msg_byte_ptr % INT_SIZE;
+    if modulo != 0 {
+        let msg_field = msg_block[int_byte_ptr];
+        for i in 0..INT_SIZE {
+            if i < INT_SIZE - modulo {
+                assert_eq(from_field_byte(msg_field, msg_byte_ptr), zero as u8);
+                msg_byte_ptr = msg_byte_ptr + 1;
+            }
+        }
+        int_byte_ptr = int_byte_ptr + 1;
+    }
+
+    // Check the rest of the numbers.
+    for i in 0..max_int_byte_ptr {
+        if i >= int_byte_ptr {
             assert_eq(msg_block[i], zero);
         }
     }
 }
 
+// Verify that up to the byte pointer the two blocks are equal.
+// At the byte pointer the new block can be partially zeroed.
+fn verify_msg_block_equals_last(
+    msg_block: MSG_BLOCK,
+    last_block: MSG_BLOCK,
+    mut msg_byte_ptr: BLOCK_BYTE_PTR,
+) {
+    // msg_byte_ptr is the position at which they are no longer have to be the same.
+    // First integer which is supposed to be (partially) zero contains that pointer.
+    let mut int_byte_ptr = msg_byte_ptr / INT_SIZE;
+
+    // Check partial zeroes.
+    let modulo = msg_byte_ptr % INT_SIZE;
+    if modulo != 0 {
+        // Reconstruct the partially zero field from the last block.
+        let last_field = last_block[int_byte_ptr];
+        let mut msg_field: u32 = 0;
+        // Reset to where they are still equal.
+        msg_byte_ptr = msg_byte_ptr - modulo;
+        for i in 0..INT_SIZE {
+            msg_field = msg_field * 256;
+            if i < modulo {
+                msg_field = msg_field + from_field_byte(last_field, msg_byte_ptr) as u32;
+                msg_byte_ptr = msg_byte_ptr + 1;
+            }
+        }
+        assert_eq(msg_block[int_byte_ptr], msg_field);
+    }
+
+    for i in 0..INT_SIZE_PTR {
+        if i < int_byte_ptr {
+            assert_eq(msg_block[i], last_block[i]);
+        }
+    }
+}
+
+// Pack the value of byte into a field of an existing message block, shifting the byte according to its position.
+fn set_msg_byte(mut msg_block: MSG_BLOCK, msg_byte_ptr: BLOCK_BYTE_PTR, msg_byte: u8) -> MSG_BLOCK {
+    let i = msg_byte_ptr / INT_SIZE;
+    msg_block[i] = set_field_byte(msg_block[i], msg_byte_ptr, msg_byte);
+    msg_block
+}
+
+// Replace one byte in a field.
+fn set_field_byte(msg_field: INT_FIELD, msg_byte_ptr: BLOCK_BYTE_PTR, msg_byte: u8) -> INT_FIELD {
```

Comment:
I think it would be worth exploring just merging `from_field_byte` and `to_field_byte` into this method. We could then merge the for-loops which would reduce some bytecode size/runtime cost

Reply:
This functions is not used any more apart from tests. Does removing it reduce the circuit or bytecode size?